src/extrema_set.rs

```diff
@@ -0,0 +1,42 @@
+/// Implementation guts for `min_set`, `min_set_by`, and `min_set_by_key`.
+pub fn min_set_impl<I, K, F, L>(mut it: I,
+                                mut key_for: F,
+                                mut lt: L) -> Option<Vec<I::Item>>
+    where I: Iterator,
+          F: FnMut(&I::Item) -> K,
+          L: FnMut(&I::Item, &I::Item, &K, &K) -> bool,
+{
+    let (mut result, mut current_key) = match it.next() {
+        None => return None,
+        Some(element) => {
+            let key = key_for(&element);
+            (vec![element], key)
+        }
+    };
+
+    for element in it {
+        let key = key_for(&element);
+        if lt(&element, &result[0], &key, &current_key) {
+            result.clear();
+            result.push(element);
+            current_key = key;
+        } else if !lt(&result[0], &element, &current_key, &key) {
+            result.push(element);
+        }
+    }
+
+    Some(result)
+}
+
+/// Implementation guts for `ax_set`, `max_set_by`, and `max_set_by_key`.
+pub fn max_set_impl<I, K, F, L>(it: I,
```

Comment:
I am not sure if I would keep `max_set_impl` or if I would instead have a single `extremal_value_set_impl` (substituting both `min_set_impl` and `max_set_impl`) that would be called with an appropriate function.


This way we would have a bit fewer customization points: We would essentially always be calling `extremal_value_set_impl`, only the functor would change. I.e. we would e.g. call `extremal_value_set_impl`
* with `x<y` in `min_set`
* with `x>y` in `max_set`

Reply:
Sure, that would work well also. As said above, I'm happy to do the change, but will not have the time to do so for a couple of weeks.